Update webpack to 4.0.0

flut1 opened the pull request:
I updated everything to make the build and dev commands work without error. However, we should thoroughly review the new configuration before merging. Particularly:

 - the mode: 'development' and mode: 'production' provide a lot of default config for us. This might allow us to remove some of our existing config.
 - `CommonsChunkPlugin` has been replaced with the `optimization` config. I am now using the default configuration for that. We should verify if that is what we want. See https://gist.github.com/sokra/1522d586b8e5c0f5072d7565c2bee693
 - Same as above for the uglify plugin. We also need to check if the default options suffice
 - Added an alpha version of `extract-text-webpack-plugin`. We might want to wait for stable
 - We should check if there are any new webpack options we want to take advantage of

flut1 commented:
There are some issues with `awesome-typescript-loader` that need to be resolved first. See s-panferov/awesome-typescript-loader#534

ThaNarie commented on build-tools/config/webpack/webpack.prod.conf.js:
```diff
@@ -62,21 +63,21 @@ const webpackConfig = merge(baseWebpackConfig, {
     new webpack.DefinePlugin({
       'process.env': env,
     }),
     new webpack.NamedChunksPlugin(),
```
This one is automatically on in `development` mode, so can be removed.

ThaNarie replied:
Ah wait, this is prod, maybe we want to leave it on here :) (but can be removed from dev config)

ThaNarie commented:
`webpack.NoEmitOnErrorsPlugin` can also be removed if i'm correct.

ThaNarie commented:
Not sure if we want this, but:

> Resolving can now be configured with module.rules[].resolve. It's merged with the global configuration.

So we could add `.vue` resolve in the vue loader config, so if you disable the loader rules, the resolving also will be removed. Best practices and all..

ThaNarie commented:
> ModuleConcatenationPlugin -> optimization.concatenateModules (on by default in production mode)

So we can remove this one as well

flut1 commented:
I implemented the suggestions by @ThaNarie  above. 
Updated awesome-typescript-loader to 5.0.0-0, which fixes most issues. However, there is still a problem with the "lib" option in `tsconfig.json`. I removed it for now, because we seem te be using the default anway

flut1 commented:
@hjeti  I fixed the TS issues (see latest commit). As far as I know this can be safely merged. It might be good though for someone to test these changes more extensively (for example, apply them in an existing project)

flut1 commented:
Just tested it by applying the changes to an existing project. All worked fine (tested dev server and prod build).
